test/lib/ansible_test/_data/sanity/pylint/plugins/deprecated.py

```diff
@@ -118,25 +154,41 @@ def visit_call(self, node):
                                 # This is likely a variable
                                 return
                             version = keyword.value.value
-                if not version:
+                        if keyword.arg == 'date':
+                            if isinstance(keyword.value.value, astroid.Name):
+                                # This is likely a variable
+                                return
+                            date = keyword.value.value
+                if not version and not date:
                     try:
                         version = node.args[1].value
                     except IndexError:
                         self.add_message('ansible-deprecated-no-version', node=node)
                         return
+                if version and date:
+                    self.add_message('ansible-deprecated-both-version-and-date', node=node)
+                    return
 
-                try:
-                    loose_version = self.version_constructor(str(version))
-                    if self.is_collection and self.collection_version is not None:
-                        if self.collection_version >= loose_version:
-                            self.add_message('collection-deprecated-version', node=node, args=(version,))
-                    if not self.is_collection and ANSIBLE_VERSION >= loose_version:
-                        self.add_message('ansible-deprecated-version', node=node, args=(version,))
-                except ValueError:
-                    if self.is_collection:
-                        self.add_message('collection-invalid-deprecated-version', node=node, args=(version,))
-                    else:
-                        self.add_message('ansible-invalid-deprecated-version', node=node, args=(version,))
+                if version:
+                    try:
+                        loose_version = self.version_constructor(str(version))
+                        if self.is_collection and self.collection_version is not None:
+                            if self.collection_version >= loose_version:
+                                self.add_message('collection-deprecated-version', node=node, args=(version,))
+                        if not self.is_collection and ANSIBLE_VERSION >= loose_version:
+                            self.add_message('ansible-deprecated-version', node=node, args=(version,))
+                    except ValueError:
+                        if self.is_collection:
+                            self.add_message('collection-invalid-deprecated-version', node=node, args=(version,))
+                        else:
+                            self.add_message('ansible-invalid-deprecated-version', node=node, args=(version,))
+
+                if date:
+                    try:
+                        if parse_isodate(date) < datetime.date.today():
+                            self.add_message('ansible-deprecated-date', node=node, args=(date,))
```

Comment:
I'm not sure if we can actually have this sanity check enabled just by checking the date. I unfortunately don't have a good alternative but this will be problematic for a few reasons.

* The dep by date is meant to enforce that deprecations will have a minimum of 2 years before they are removed, it's not to say that this will be removed at exactly 2 years time
* To remove a feature we would typically need to increase the major version of the collection, doing this at random times just because a dep period has been reached won't paint the best picture of stability of the collection

In short having a sanity test fail when we don't have any immediate plans to actually remove the feature means either we need to ignore the error and defeat the purpose of the check or just live with the error, making CI failures harder to diagnose.

An example scenario of this problem would be the maintenance releases of a certified collection. These collections are currently set to have a rolling release and a 24 month maintenance release. The latter is a bug fix only release with no features being added or removed so it could potentially have a deprecated feature 6 months into it's lifecycle and we need to ignore this sanity check for that branch entirely.

Another scenario is for the rolling release, say we have 2 dep by dates set

1. module x has a `removed_at_date: 2020-06-01`
2. module y has a `removed_at_date: 2020-08-01`

In this scenario we would effectively have to remove the feature at the start of June **and** August to silence the ansible-test sanity checks and in the meantime we either release 2 major bumps in that timeframe or wait the 2 months before doing the next release.

Maybe the best option here is to have this as a manual check, i.e. it's run before creating the release to ensure that all deprecations that can be removed are removed rather than something done in CI.

Reply:
This test is part of the optional tests, i.e. you explicitly have to run `ansible-test sanity` with `--enable-optional-errors` for it to be a (reported) error. So it will only have an impact if people explicitly request this. (This might change in the future to a non-failing warning, but the result will still be the same: this won't make tests fail from a date on until these tests are explicitly requested. This definitely won't happen for 2.10 though.)

Reply:
(Optional errors were introduced in #66920, though not used there yet. This would be the first use of that feature.)

Reply:
Ah that's fantastic, solves this problem quite nicely. Having a quick skim of #66920 it seems like `--enable-optional-errors` is a true/false switch so by setting it you would enable all the optional errors. Is it mean to work in conjuction with the test name like `ansible-test sanity --test ansible-deprecated-date --enable-optional-errors` or is that not possible right now?

Reply:
`--test` only works with tests like `validate-modules` or `pylint`, not with explicit error codes returned by one of the tests. So you can only enable all of them, or none. @mattclay and me were discussing about that last week, I think the plan is to improve the interface eventually, but not for 2.10. Since it would probably be the only optional error for 2.10 it should be ok for now :)

Reply:
Fair enough.